### ISSUE TITLE
Links are silently removed when deleting a node

### ISSUE THREAD

**Issue by samhatchett** [OpenWaterAnalytics/EPANET#473]
https://github.com/OpenWaterAnalytics/EPANET/blob/61847dbcbc2dbfd82eb3897977043add5cd03ebd/src/epanet.c#L1926

While the silent deletion of links is canonical from the perspective of the time-tested UI, this has some unintended consequences when the toolkit is used from within an editing application - especially an application with an undo stack. To have the data model change in this way without some kind of notification is a general anti-pattern. The toolkit could assume that a client application will adequately interrogate the data model before making such a modification, but that's presuming a lot and seems fragile.

I would advocate for handling this the same way as if a control references the node. The API user has the option to "force" the deletion (as `UNCONDITIONAL`), or do a "safe" deletion meaning that the toolkit won't auto-delete links and controls that depend on that node. Discussion of that occurred here: https://github.com/OpenWaterAnalytics/EPANET/issues/240
And more conversation in the PR for handling rules, which was merged here: https://github.com/OpenWaterAnalytics/EPANET/pull/329

Any strong opinions? I am more than happy to contribute a PR depending on feedback.

**Comment by LRossman** [comment]
If you don't delete the link when one of its nodes is deleted then that end of the link, stored as a node index (I know indexes are bad 😉), will attach the link to a node whose index was originally one higher than the index of the deleted node, thus breaking the integrity of the network's structure. How do the various graph libraries (like Boost Graph) handle node/vertex deletion?

**Comment by samhatchett** [comment]
To be clear, I'm suggesting that a node be "un-delete-able" if it is connected to a link, unless you pass in the "unconditional", in which case the link would also be deleted. I agree that it does not make sense to have a link with one un-attached end. BGL is great for graph abstractions, but as a generic graph library it doesn't care about the domain we are interested in.

**Comment by LRossman** [comment]
OK, I see now. So for "conditional" deletion the node is deleted only if all of its links have been explicitly deleted beforehand. Sounds good to me. To implement this change I think you only have to modify the following section of `EN_deletenode()` in _epanet.c_ as follows:
```
    if (actionCode == EN_CONDITIONAL)
    {
        if (incontrols(p, NODE, index)) return 261;
        for (i = 1; i <= net->Nlinks; i++)
        {
            if (net->Link[i].N1 == index ||
                net->Link[i].N2 == index)  return 259;
        }
    }
```

A new error code, 259, would have to be added to errors.dat that might look as follows:
```
DAT(259,"attempt to delete a node that still has links connected to it")
```

**Comment by michaeltryby** [comment]
Question? What kind of mechanism would be useful for a client to adequately “interrogate” the data model. What information beyond an elements existence or lack there of would it require?

**Comment by LRossman** [comment]
Regarding model interrogation, the toolkit has no mechanism to inform a client of which links are connected to a given node nor which controls refer to it. Hence if conditional node deletion fails it only tells a toolkit user that the node is used by other objects, but not which ones. They would have to write code that searches the entire set of links and controls looking for where the node appears (as in the code snippet shown above) and take the appropriate action to remove the reference to it.

**Comment by michaeltryby** [comment]
Has no mechanism yet ...

**Comment by samhatchett** [comment]
right, convenience wrappers for model interrogation are not a bad idea, and pretty easy to implement. Future convenience aside, do I interpret correctly that there is good agreement about this proposed change (namely, refusing to delete a node that has links attached)?

**Comment by LRossman** [comment]
I'm in agreement.

**Comment by michaeltryby** [comment]
I agree too

**Comment by eladsal** [comment]
Closed by #478
